firebase/functions/src/index.ts

```diff
@@ -0,0 +1,532 @@
+    } catch (e) {
+        if (e instanceof createError.HttpError) {
+            throw e;
+        } else {
+            console.error('Caught error looking up deck', e)
+            throw createError(500);
+        }
+        
+    }
+    if (!deckDoc.exists) {
+        throw createError(400, null, {
+            headers: {
+                'X-Reason': 'The deck does not exist'
+            }
+        });
+    }
+    
+
+    const deck = deckDoc.data();
+    const gameDocRef = firestore.collection('Games').doc(gameId);
+    let gameDoc;
+    try {
+        gameDoc = await gameDocRef.get();
+    } catch (e) {
+        console.error('Caught error getting game document', e)
+        throw createError(500);
+    }
+    if (!gameDoc.exists) {
+        throw createError(400, null, {
+            headers: {
+                'X-Reason': 'The target game does not exist'
+            }
+        });
+    }
+    
+    return createPlayerInFirestore(gameDocRef, player, deck);
+}
+
+export const joinGame = functions.https.onRequest((request, response) => {
+    if (request.method !== 'POST') {
+        response.status(405)
+            .header('Allow', 'POST')
+            .send('Method Not Allowed');
+    }
+    
+    const gameId = request.body.game;
+    const player = request.body.player;
+    const deckId = request.body.deck;
+
+    if (gameId === undefined) {
+        response
+            .status(400)
+            .send('Body must include "game" attribute denoting target game\'s ID');
+    }
+    if (player === undefined) {
+        response
+            .status(400)
+            .send('Body must include "player" attribute denoting host player\'s username');
+    }
+    if (deckId === undefined) {
+        response
+            .status(400)
+            .send('Body must include "deck" attribute denoting host player\'s desired deck ID');
+    }
+
+    console.log('Joining existing game (' + gameId + ') with hosting player ' + player);
+    console.log('Using deck with ID ' + deckId);
+
+    joinGameHelper(gameId, player, deckId)
+        .then(() => {
+            response.status(200)
+                .send();
+        })
+        .catch((error) => {
+            errorResponseMapper(error, response);
+        })
+});
+
+const determineTurnOrderHelper = async function(gameId: string): Promise<void> {
+    try {
+        const gameDocRef = firestore.collection('Games').doc(gameId);
+        const gameDoc = await gameDocRef.get();
+        if (gameDoc.exists) {
+            const playersCol = await gameDocRef.collection('Players').get();
+            const playerOrder = playersCol.docs
+                .map((doc) => doc.id)
+                .sort(() => uuidv4());
+            
+            const turnOrder = {};
+            // This is just some crap way to turning the array into a map (even though
+            // the keys are just the position and the values are the same)
+            for (let i = 0; i < playerOrder.length; i++) {
```

Comment:
Is `turnOrder` supposed to look something like this?
```
{
  1: Andrew,
  2: Eric,
  3: Anthony
}
```
If so, it might be cleaner to write it like this (assuming `playerOrder` is an array of names `[Andrew, Eric, Anthony]`:
```
const turnOrder = playerOrder.reduce((order, player, index) => ({
  ...order,
  [index]: player,
}), {});
```
`reduce()` takes in an array and returns a single object. `...` is the spread operator, used above to take all the items currently in `order` and add it to this new object we're returning.

Bonus lodash:
```
const turnOrder = _.reduce(playerOrder, (order, player, index) => ({
  ...order,
  [index]: player,
}), {});
```

Reply:
Actually, messing around with it a little more,
```
const turnOrder = {...playerOrder};
```
might get the job done just as well.

Reply:
I love the spread operator. Easily one of my favorite things in ES6